### PR TITLE
Add a DoH implementation for Intra

### PR DESCRIPTION
**Pull request by bemasc**

### CODE REVIEW

File: tunnel/intra/doh.go [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+		fmt.Printf("Query forwarding failed: %v\n", err)
+		c.Close()
+	}
+}
+
+// Accept a DNS-over-TCP socket from a stub resolver, and connect the socket
+// to this DNSTransport.
+func Accept(t DNSTransport, c io.ReadWriteCloser) {
+	qlbuf := make([]byte, 2)
+	for n, err := c.Read(qlbuf); err == nil && n == 2; n, err = c.Read(qlbuf) {
```
Comment (alalamav) [inline tunnel/intra/doh.go]: Do we want to log query lenght read failures? Can we, using this construct?

File: tunnel/intra/udp.go [hunk 2]
```diff
@@ -147,6 +155,17 @@ func (h *udpHandler) Connect(conn core.UDPConn, target *net.UDPAddr) error {
 	return nil
 }
 
+func (h *udpHandler) doDoh(dns DNSTransport, t *tracker, conn core.UDPConn, data []byte) {
+	resp, err := dns.Query(data)
```
Comment (alalamav) [inline tunnel/intra/udp.go]: Do we want to log the error?

File: tunnel/intra/doh.go [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+	DNSTransport
+	url      string
+	port     int
+	ips      ipmap.IPMap
+	client   http.Client
+	listener DNSListener
+}
+
+// Wait up to three seconds for the TCP handshake to complete.
+const tcpTimeout time.Duration = 3e9
```
[the remark this answers is not in the document]
Comment (bemasc) [reply tunnel/intra/doh.go]: Thanks for the reminder.  Done.

File: tunnel/intra/doh.go [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+
+func (t *transport) dial(network, addr string) (net.Conn, error) {
+	fmt.Printf("Dialing %s\n", addr)
```
[the remark this answers is not in the document]
Comment (bemasc) [reply tunnel/intra/doh.go]: Good idea.  Done throughout.

File: tunnel/intra/doh.go [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+		return nil, err
+	}
+	if parsedurl.Scheme != "https" {
+		return nil, fmt.Errorf("Bad scheme: %s", parsedurl.Scheme)
+	}
+	// Resolve the hostname and put those addresses first.
+	portStr := parsedurl.Port()
+	var port int
+	if len(portStr) > 0 {
+		port, err = strconv.Atoi(parsedurl.Port())
```
[the remark this answers is not in the document]
Comment (bemasc) [reply tunnel/intra/doh.go]: Done

File: tunnel/intra/doh.go (hunk 1 above)
Comment (bemasc) [reply tunnel/intra/doh.go]: OK, done.

File: tunnel/intra/udp.go (hunk 2 above)
Comment (bemasc) [reply tunnel/intra/udp.go]: Good idea.  Done.

**Pull request merged**